3.32.0: Adds new SDK version and contract files (#3687)
Azure/azure-cosmos-dotnet-v3: changelog.md

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

### <a name="3.32.0"/> [3.32.0](https://www.nuget.org/packages/Microsoft.Azure.Cosmos/3.32.0) - 2023-02-03
#### Fixed
- [#3632](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3632) Query: Fixes handling of CosmosUndefined, CosmosGuid and CosmosBinary in unordered DISTINCT

> - [#3640](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3640) Upgrade Resiliency: Refactors Implementation for Opening Rntbd Connections to Backend Replicas in Direct Mode.

Reviewer: @kundadebdatta Is this something that is released or under the env variable? If under an env variable, then we should remove it?

Reviewer 2: @ealsur : This was merged (however not released yet) back in October with this [PR 3508](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3508) and this new PR - 3640 just refactors the implementation to match with the Direct code refactors. The replica validation feature changes that is backed by the environment variable is yet to be merged.

Reviewer: Then this PR is not really shipping anything yet, correct? We can remove it from the changelog?

Reviewer 2: This PR refactors the `CosmosClient.CreateAndInitializeAsync()` to initialize and warm up the caches by opening up the Rntbd connections to the backend replica nodes, instead of using the dummy querys on all feed ranges to open connection. This is separate from the replica validation feature _(which internally uses some of the changes from this PR to open connections to the replica nodes)_, and IMO should be included as a part of the release.

Reviewer: What about https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3508? Should we then mention both and rewrite the text to say something about CreateAndInitializeAsync? The current text is not very clear

Reviewer 2: That's probably a good idea! Let's mention both and update the verbiage. Let me know how you feel about the below text for #3640 :

> Upgrade Resiliency: Refactors `CosmosClient` Initialization to use `IOpenConnectionHandler` for Opening Rntbd Connections to Backend Replicas in Direct Mode.

@aavasthy - If we agree, can you please update the verbiage ?

Author: updated

> - [#3640](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3640) Fix for service using Microsoft.Azure.Cosmos 3.30.1 getting token expiry errors

Reviewer: This PR link is repeated, could it be a different one?

Author: This PR closed two different issues.

Reviewer: You are correct

- [#3645](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3645) Change Feed Processor: Fixes behavior with StartTime on Local
- [#3656](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3656) Region Availability: Adds Poland Central Region For Public Usage.
- [#3636](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3636) CosmosClientOptions: Adds ServerCertificateCustomValidationCallback for Http and TCP

> - [#3664](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3664) Query: Adds EnableOptimisticDirectExecution flag to QueryRequestOptions

Reviewer: Remove, this was incorrectly tagged as public (missing [Internal]), the API that was added was not public

Author: Removed PR 3664

> - [#3680](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3680) LocalQuorum: Adds Quorum reads on Consistent Prefix Accounts

Reviewer: Missing:

* https://github.com/Azure/azure-cosmos-dotnet-v3/commit/612337cfbd445a95042457b672819e36e0a6319e
* https://github.com/Azure/azure-cosmos-dotnet-v3/commit/6632168410079b849aa5d6f31f9bdd6f1ef52da1
* https://github.com/Azure/azure-cosmos-dotnet-v3/commit/b24eeb96babef246d8deb2305e08accb7ceffc7d
* https://github.com/Azure/azure-cosmos-dotnet-v3/commit/010e4ba7198b62cb13b680baf4f6b2d6f9538cef
* https://github.com/Azure/azure-cosmos-dotnet-v3/commit/2681be80b9674fc8a09f1a7552894ef8741b23e2
* https://github.com/Azure/azure-cosmos-dotnet-v3/commit/c38346dd8d029bf32dc523c678b6df95e8c1b9b9
* https://github.com/Azure/azure-cosmos-dotnet-v3/commit/ab1f249c4074545765e466ece547a8c96fe51095
* https://github.com/Azure/azure-cosmos-dotnet-v3/commit/68b980523bd89fa4b298b1589e6d7d5148df3748
* https://github.com/Azure/azure-cosmos-dotnet-v3/commit/c0ac87af429fcf526b125f8777b02a3bd3997704
* https://github.com/Azure/azure-cosmos-dotnet-v3/commit/63a846de8cb108cccb17addefa9b1937557979ca
* https://github.com/Azure/azure-cosmos-dotnet-v3/commit/748f86c052934796590c12d80e836f7f031d8525
* https://github.com/Azure/azure-cosmos-dotnet-v3/commit/f679c3ba9368b3b297be873e41c131729da123fd
* https://github.com/Azure/azure-cosmos-dotnet-v3/commit/ff01c1448fe40226d2a8433c135e48d47f66d9b9

Author: updated

### <a name="3.32.0-preview"/> [3.32.0-preview](https://www.nuget.org/packages/Microsoft.Azure.Cosmos/3.32.0-preview) - 2023-02-03

#### Added
- [#3596](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3596) Full Fidelity Change Feed: Adds new LatestVersion and AllVersionsAndDeletes to ChangeFeedMode

> - [#3598](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3598) AI integration: Adds IsDistributedTracingEnabled flag as public API to enable/disable this feature

Reviewer: @sourabh1007 Is this the right text? Can we better describe the introduction of this as a feature. Maybe something like:

```suggestion
- [#3598](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3598) AI integration: Adds Distributed Tracing support. Enabled by default, which can be disabled through `CosmosClientOptions.IsDistributedTracingEnabled`
```

Author: @sourabh1007 updated the title based on the above suggestion. Let me know if you're good with the change.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

### <a name="3.32.0"/> [3.32.0](https://www.nuget.org/packages/Microsoft.Azure.Cosmos/3.32.0) - 2023-02-03
#### Fixed
- [#3466](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3466) ClientRetryPolicy: Fixes behavior to Meta-data write operations in multimaster accounts
- [#3498](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3498) PartitionKey: Fixes NullRef in toString handling for None for PartitionKey.ToString()
- [#3385](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3385) Query: Fixes LINQ ToString got absorbed during translation
- [#3406](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3406) Query: Fixes LINQ to use custom serializer to serialize constant values in Query

> - [#3496](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3496) ReadThroughputAsync: Adds XML comment to Database.ReadThroughputAsync definition

Reviewer: ```suggestion
- [#3496](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3496) Documentation: Adds XML comment to Database.ReadThroughputAsync definition
```

> - [#3508](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3508) CosmosClient Initialization: Adds implementation for opening Rntbd connections to backend replica nodes in Direct mode

Reviewer: ```suggestion
- [#3508](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3508) CosmosClient Initialization: Fixes implementation for opening Rntbd connections to backend replica nodes in Direct mode
```

> - [#3508](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3508) CosmosClient Initialization: Adds implementation for opening Rntbd connections to backend replica nodes in Direct mode

Reviewer: ```suggestion
- [#3508](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3508) CosmosClient Initialization: Refactors implementation for opening Rntbd connections to backend replica nodes in Direct mode
```

- [#3519](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3519) Diagnostics: Removes unused properties and reduces size 
- [#3495](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3495) Query: Fixes partition range evaluation for spatial queries

> - [#3566](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3566) Change Feed Processor: Adds support for Resource Tokens

Reviewer: Anything that is "Adds" should be in an "Added" section on the same level as the "Fixed" section

Author: Even if it is a bug fix?

Reviewer: If its a bug fix, the text should be Fixes, right?

- [#3399](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3399) Query: Fixes default to BadRequestException in case of internal errors in ServiceInterop

> - [#3555](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3555) Client Retry Policy: Adds HTTP timeouts with request-level cross-region retry

Reviewer: ```suggestion
- [#3555](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3555) Availability: Adds HTTP timeouts with request-level cross-region retry
```

- [#3574](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3574) Query: Fixes incorrect FeedResponse.Count when result contains undefined elements
- [#3577](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3577) Trace: Fixes Tracing/diagnostics hour-times to 24Hours

> - [#3581](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3581) AI Integration: Removed EnableDiagnosticsTraceForAllRequests Config

Reviewer: Remove

- [#3509](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3509) Query: Adds ALL Scalar Expression 
- [#3632](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3632) Query: Fixes handling of CosmosUndefined, CosmosGuid and CosmosBinary in unordered DISTINCT

> - [#3640](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3640) Upgrade Resiliency: Refactors CosmosClient.CreateAndInitializeAsync() to use IOpenConnectionHandler for Opening Rntbd Connections to Backend Replicas in Direct Mode.

Reviewer: Remove. In the line where we have `[#3508](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3508)` add this PR also, so it ends up being `- [#3508](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3508), [#3640](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3640)`

> - [#3640](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3640) Fix for service using Microsoft.Azure.Cosmos 3.30.1 getting token expiry errors

Reviewer: ```suggestion
- [#3640](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3640) Token expiration: Fixes token expired errors happening on some environments
```

- [#3645](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3645) Change Feed Processor: Fixes behavior with StartTime on Local
- [#3656](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3656) Region Availability: Adds Poland Central Region For Public Usage.
- [#3636](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3636) CosmosClientOptions: Adds ServerCertificateCustomValidationCallback for Http and TCP

> - [#3680](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3680) LocalQuorum: Adds Quorum reads on Consistent Prefix Accounts

Reviewer: This is internal, remove

> - [#3643](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3643) Fixed CosmosClientBuilder.WithConnectionModeGateway documentation

Reviewer: ```suggestion
- [#3643](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3643) Documentation: Fixed CosmosClientBuilder.WithConnectionModeGateway documentation
```

- [#3643](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3643) Documentation: Fixed CosmosClientBuilder.WithConnectionModeGateway documentation
- [#3579](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3579) Documentation: Fixes EUAP in Comments

#### Added
- [#3566](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3566) Change Feed Processor: Adds support for Resource Tokens
- [#3555](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3555) Availability: Adds HTTP timeouts with request-level cross-region retry
- [#3509](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3509) Query: Adds ALL Scalar Expression 
- [#3656](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3656) Region Availability: Adds Poland Central Region For Public Usage.
- [#3636](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3636) CosmosClientOptions: Adds ServerCertificateCustomValidationCallback for Http and TCP

### <a name="3.32.0-preview"/> [3.32.0-preview](https://www.nuget.org/packages/Microsoft.Azure.Cosmos/3.32.0-preview) - 2023-02-03

#### Added

> - [#3596](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3596) Full Fidelity Change Feed: Adds new LatestVersion and AllVersionsAndDeletes to ChangeFeedMode

Reviewer 3: How about calling the older name removal?

Reviewer: Maybe something like: 

```suggestion
- [#3596](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/3596) Full Fidelity Change Feed: Adds new LatestVersion to ChangeFeedMode. `FullFidelity` is now renamed to `AllVersionsAndDeletes`.
```
